File: src/SubPals.py
def checkPals(s: str, subs: list[str]) -> list[list[int]]:
    base = 31
    mod = int(1e9)+7
    pows = [0]*(len(s)+1)
    pows[0] = 1
    for i in range(1, len(pows)):
        pows[i] = (pows[i-1]*base)%mod

    # Hashcodes of all elements of sub, separated by parity
    hs = [set(), set()]
    for sub in subs:
        hashCode = 0
        for i in range(int(len(sub)/2), len(sub)):
            charVal = ord(sub[i])-ord('a')+1
            hashCode = (hashCode*base+charVal)%mod
        hs[len(sub)&1].add(hashCode)

    # Insert a space between all characters and a special character at front and end
    # Example: "abc" -> "< a b c >"
    arr = [0] * (2*len(s)+3)
    for i in range(len(s)):
        arr[2*i+2] = ord(s[i])-ord('a')+1
    arr[0] = -1
    arr[len(arr)-1] = -2

    # Length of longest (not necessarily valid) palindrome centered at i
    dp = [0] * (2*len(s)+3)

    # Hashcode of longest (not necessarily valid) palindrome centered at i
    codes = [0] * (2*len(s)+3)

    # Length of first invalid palindrome centered at i
    bounds = [(2*len(s)+3) for _ in range((2*len(s)+3))]
    left, right = 0, 0

    # Running hashcode of interval [i, right]
    hashCode = 0

    # Manacher's algorithm with the codes and bounds arrays piggybacking
    # off of the string traversal/mirroring
    for i in range(2, len(dp)-2):
        if (i+1 >= right):
            codes[i] = arr[i]
        elif (right-i < dp[right-i+left]):
            dp[i] = right-i
            codes[i] = hashCode if arr[i] == 0 else hashCode*base+arr[i]
            if (right-i > bounds[right-i+left]): 
                bounds[i] = bounds[right-i+left]
        else:
            dp[i] = dp[right-i+left]
            codes[i] = codes[right-i+left]
            bounds[i] = bounds[right-i+left]

        while arr[i - dp[i] - 1] == arr[i + dp[i] + 1]:
            if (arr[i - dp[i] - 1] != 0): 
                codes[i] = (codes[i] * base + arr[i - dp[i] - 1]) % mod

            # If current hashcode contained in subs, current palindrome is invalid
            if (codes[i] in hs[1^i&1]): 
                bounds[i] = min(bounds[i], dp[i])

            dp[i] += 1


        if i+dp[i] > right:
            right = i+dp[i]
            left = 2*i-right
            hashCode = codes[i]

        if (arr[i] != 0):
            hashCode = (hashCode-arr[i]*pows[int((right-i)/2)]+base*mod)%mod

    # Rightmost index whose inclusion would necessarily mean containing an invalid palindrome
    bound = 0
    for i in range(2, len(dp)-2):
        bound = max(bound, i - bounds[i])
        dp[i] = min(dp[i], i - bound)

    ans = [[dp[i*2+2], dp[i*2+3]] for i in range(len(s))]

    return ans

File: src/SubPals.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <set>
#include <iostream>
#include <algorithm>
using namespace std;

vector<vector<long>> checkPals(
string s,
vector<string> &subs)
{
    long base = 31;
    long long mod = (long long)1E9 + 7;
    vector<long long> pows(s.length() + 1);
    pows[0] = 1;
    for (long i = 1; i < pows.size(); i++)
    {
        pows[i] = (pows[i - 1] * base) % mod;
    }
    vector<set<long long>> hs(2);
    for (long i = 0; i < hs.size(); i++)
    {
        hs[i] = set<long long>();
    }
    for (string str : subs)
    {
        long hashCode = 0;
        for (long i = long(str.length() / 2); i < str.length(); i++)
        {
            long charVal = str[i] - 'a' + 1;
            hashCode = (hashCode * base + charVal) % mod;
        }
        hs[str.length() & 1].insert(hashCode);
    }
    vector<long> arr(2 * s.length() + 3);
    for (long i = 0; i < s.length(); i++)
    {
        arr[2 * i + 2] = s[i] - 'a' + 1;
    }
    arr[0] = -1;
    arr[arr.size() - 1] = -2;
    vector<long> dp(2 * s.length() + 3);
    vector<long long> codes(2 * s.length() + 3);
    vector<long> bounds(2 * s.length() + 3);
    fill(bounds.begin(),bounds.end(),bounds.size());
    long left = 0;
    long right = 0;
    long hashCode = 0;
    for (long i = 2; i < dp.size() - 2; i++)
    {
        if (i + 1 >= right)
        {
            codes[i] = arr[i];
        }
        else if (right - i < dp[right - i + left])
        {
            dp[i] = right - i;
            codes[i] = ((arr[i] == 0) ? hashCode : hashCode * base + arr[i]);
            if (right - i > bounds[right - i + left])
            {
                bounds[i] = bounds[right - i + left];
            }
        }
        else
        {
            dp[i] = dp[right - i + left];
            codes[i] = codes[right - i + left];
            bounds[i] = bounds[right - i + left];
        }
        for (; arr[i - dp[i] - 1] == arr[i + dp[i] + 1]; dp[i]++)
        {
            if (arr[i - dp[i] - 1] != 0)
            {
                codes[i] = (codes[i] * base + arr[i - dp[i] - 1]) % mod;
            }
            if ((hs[1 ^ i & 1].find(codes[i]) != hs[1 ^ i & 1].end()))
            {
                bounds[i] = min(bounds[i],dp[i]);
            }
        }
        if (i + dp[i] > right)
        {
            right = i + dp[i];
            left = 2 * i - right;
            hashCode = codes[i];
        }
        if (arr[i] != 0)
        {
            hashCode = (hashCode - arr[i] * pows[long((right - i) / 2)] + base * mod) % mod;
        }
    }
    long bound = 0;
    for (long i = 2; i < dp.size() - 2; i++)
    {
        bound = max(bound,i - bounds[i]);
        dp[i] = min(dp[i],i - bound);
    }
    vector<vector<long>> ans(s.length(),vector<long>(2));
    for (long i = 0; i < s.length(); i++)
    {
        ans[i] = vector<long>{dp[i * 2 + 2], dp[i * 2 + 3]};
    }
    return ans;
}



int main() {
    /* Enter your code here. Read input from STDIN. Prlong output to STDOUT */   

    string s;
    vector<string> subs;
    long n;

    cin >> s;
    cin >> n;

    string input;
    for (long i = 0; i < n; i++) {
        cin >> input;
        subs.push_back(input);
    }

    vector<vector<long>> ans = checkPals(s, subs);

    for (long i = 0; i < ans.size(); i++) {
        cout << ans[i][0] << " " << ans[i][1] << "\n";
    }

    return 0;
}